src/limb.rs

```diff
@@ -51,17 +49,62 @@ pub const LIMB_BYTES: usize = (LIMB_BITS + 7) / 8;
 /// References a positive integer range `[1..max_exclusive)`.
 /// `max_exclusive` is assumed to be public, not secret.
 pub struct Range<'a> {
-    pub max_exclusive: &'a [Limb],
+    max_exclusive: &'a [Limb],
+    sampling_reduce_when_over_bound: bool,
+    sampling_most_sig_limb_mask: Limb,
 }
 
 impl <'a> Range<'a> {
-    pub fn from_max_exclusive(max_exclusive: &[Limb]) -> Range {
+    //TODO(DJ) Either we should stop using Range where this is used in EC code,
+    // or we should convert that code to use Range for sampling/keygen too.
+    pub fn from_max_exclusive_without_leading_zeros(max_exclusive: &[Limb]) -> Range {
+        debug_assert!(max_exclusive[max_exclusive.len()-1].leading_zeros() == 0);
+
+        Range {
+            max_exclusive: max_exclusive,
+            sampling_reduce_when_over_bound: false,
+            sampling_most_sig_limb_mask: Limb::max_value(),
+        }
+    }
+
+    pub fn from_arbitrary_max_exclusive(max_exclusive: &[Limb]) -> Range {
         debug_assert_eq!(limbs_are_zero_constant_time(max_exclusive),
                          LimbMask::False);
-        debug_assert!(max_exclusive[max_exclusive.len() - 1] > 0);
+
+        let most_significant_limb = max_exclusive[max_exclusive.len() - 1];
+
+        debug_assert!(most_significant_limb > 0);
+
+        // TODO(DJ) Eliminate `as`
+        let leading_zeros = most_significant_limb.leading_zeros() as usize;
+        if leading_zeros == 0 {
+            // If there are no leading zeros, we'll never use the special
+            // `0b100...` behavior, because it would require us to allocate
+            // an extra limb into which to sample.
```

Comment:
This seems problematic to me. Isn't it going to be the case usually that there will be zero leading zeros, because the modulus will be 2048 bits, which is the most common length for a public RSA modulus?

Maybe we should instead generate the random value into a stack-allocated buffer that has the extra space, and then copy it over?